torch/distributed/nn/api/remote_module.py

```diff
@@ -0,0 +1,245 @@
+
+    Returns:
+        A remote module instance which wraps the :class:`~nn.Module` created by the
+        user-provided ``module_creator``, it has a blocking ``forward`` method and an asynchronous
+        ``forward`` that returns a future of the ``forward`` call on the user created
+        remote module.
+
+    Example::
+        Run the following code in two different processes:
+
+        >>> # On worker 0:
+        >>> import torch
+        >>> import torch.distributed.rpc as rpc
+        >>> from torch import nn, Tensor
+        >>> from torch.distributed.nn.api.remote_module import RemoteModule
+        >>>
+        >>> @torch.jit.interface
+        >>> class LinearIf:
+        >>>     def forward(self, input: Tensor):
+        >>>         pass
+        >>>
+        >>> rpc.init_rpc("worker0", rank=0, world_size=2)
+        >>> remote_linear_module = RemoteModule(
+        >>>     "worker1", nn.Linear, args=(20, 30), module_interface_cls=LinearIf
+        >>> )
+        >>> input = torch.randn(128, 20)
+        >>> ret_fut = remote_linear_module.forward_async(input)
+        >>> ret = ret_fut.wait()
+        >>> rpc.shutdown()
+
+        >>> # On worker 1:
+        >>> import torch
+        >>> import torch.distributed.rpc as rpc
+        >>>
+        >>> rpc.init_rpc("worker1", rank=1, world_size=2)
+        >>> rpc.shutdown()
+    """
+    # Sanity checks.
+    assert rpc._is_current_rpc_agent_set(), "RemoteModule only works in RPC."
+
+    # Default arguments preperation.
+    args = args if args is not None else ()
+    kwargs = kwargs if kwargs is not None else {}
+    global_unique_name = (  # Assign a global name for the module to be created.
+        global_unique_name
+        if global_unique_name is not None
+        else _gen_global_unique_name()
+    )
+
+    if module_interface_cls is not None:
+        is_scriptable = True
+
+        # Instantiate template on remote side.
+        fut = rpc.rpc_async(
+            to, _instantiate_template, (module_interface_cls,)
+        )
+
+        # Instantiate template on local side.
+        generated_module = instantiator.instantiate_scriptable_remote_module_template(module_interface_cls)
+        remote_module_cls = generated_module._RemoteModule
+
+        # Create the module on the remote side.
+        fut.wait()  # Ensure remote_module_cls is available on remote side.
+        module_rref = rpc.rpc_sync(
+            to,
+            _scripted_module_creator_wrapper,
+            (module_creator, module_interface_cls, args, kwargs),
+        )
+    else:
+        is_scriptable = False
+        remote_module_cls = _NON_SCRIPTABLE_REMOTE_MODULE_CLS
+
+        # Create the module on the remote side.
+        module_rref = rpc.remote(to, module_creator, args, kwargs)
+
+    # Create remote_module_cls instance on local side.
+    remote_module = remote_module_cls(module_rref, is_scriptable, global_unique_name)
+
+    return remote_module
+
+
+def RemoteModule(
```

Comment:
I recall there was a discussion regarding whether this should be a `class` and annotate it's `__new__` method. Is it because that didn't work, so that we still make this a function?

Reply:
Going with this idea instead. https://github.com/pytorch/pytorch/pull/37139#discussion_r434201926